### PR TITLE
[NCL-8685] More strict url validation for gitlab

### PR DESCRIPTION
**Pull request by thescouser89**
If we are using gitlab as the internal url during build-config creation, we need to enforce scp format for the internal url for consistency.

Special thanks to @DnsZhou for his huge help on this one!

### CODE REVIEW

File: ui/app/common/directives/pncScmValidator.js [hunk 1]
```diff
@@ -54,7 +66,15 @@
...
-        return pattern.test(url);
+        if (pattern.test(url)) {
+          if (isRepoInternal(url) && isGitlabAndNotScpFormat(url)) {
+            return false;
+          } else {
+            return true;
+          }
+        } else {
+         return false;
+        }
```
Comment (DnsZhou) [inline ui/app/common/directives/pncScmValidator.js:77]: can be simplified by
`return pattern.test(url) && !(isRepoInternal(url) && isGitlabAndNotScpFormat(url));`

**Review by DnsZhou — CHANGES_REQUESTED** [review]
Hi Dustin, Thank you for working on the UI code. There are several points to simplify the code, but Functionally it works good! feel free to reach me for any question.

File: ui/app/common/directives/pncScmValidator.js [hunk 2]
```diff
@@ -42,7 +42,19 @@
    * <input pnc-scm-validator="git|ssh|http|https|git+ssh|git@" pnc-exact-host="user@host.xz:port" ... >
    * @author Martin Kelnar
    */
-  module.directive('pncScmValidator', function () {
+  module.directive('pncScmValidator', ['pncProperties', function (pncProperties) {
+
+      function isRepoInternal(url) {
+        return url.includes(pncProperties.internalScmAuthority);
```
[the remark this answers is not in the document]
Comment (thescouser89) [reply ui/app/common/directives/pncScmValidator.js:48]: It's already there: https://github.com/project-ncl/pnc-web-ui-angularjs/blob/106d372d3c85dfcf9bb4d4e5327312842a076490/ui/Gruntfile.js#L123

or am I misunderstanding something?

File: ui/app/common/directives/pncScmValidator.js [hunk 3]
```diff
@@ -42,7 +42,19 @@
...
+      }
+
+      // From NCL-8685: for gitlab internal url, only support scp format
+      function isGitlabAndNotScpFormat(url) {
+        var isGitlab = url.includes('gitlab');
```
[the remark this answers is not in the document]
Comment (thescouser89) [reply ui/app/common/directives/pncScmValidator.js:53]: Thanks!

File: ui/app/common/directives/pncScmValidator.js [hunk 4]
```diff
@@ -42,7 +42,19 @@
...
+        // scp format is: <user>@<host>:<path>
+        var patternScpFormat = new RegExp('^.+@.+:.+$');
```
[the remark this answers is not in the document]
Comment (thescouser89) [reply ui/app/common/directives/pncScmValidator.js:55]: Thanks!

File: ui/app/common/directives/pncScmValidator.js (hunk 1 above)
Comment (thescouser89) [reply ui/app/common/directives/pncScmValidator.js:77]: True! Thanks for pointing this out!

**Comment by thescouser89** [comment]
Ready for review after comment fixes.

**Review by DnsZhou — APPROVED** [review]
lgtm now!

**Pull request merged**